bindgen: systematic divergence between Lisette semantic validation and Golang compilation

I have attached a reduction created as a smoke test for `lis add`, `bindgen` and third-party Golang package interop.
`cobra` is used in this case.

Interop is a convenience here rather than a blocking requirement.

`lis build` does not complain about the code but `go build` raises the error:
`./main.go:37:30: cannot use lisette.MakeOptionSome(cobra.ShellCompDirectiveDefault) (value of struct type lisette.Option[cobra.ShellCompDirective]) as *cobra.ShellCompDirective value in struct literal`

I did not have time to investigate properly, but the problem seemed to be a collision between reference and value collision.

A similar smoke test using pgx raising another typedef mismatch, again at the bindgen boundary.

The single main.lis application file  for cobra is zipped - GitHub does not like raw Lisette attachments :-(
[main.lis.zip](https://github.com/user-attachments/files/27094398/main.lis.zip)

fwiw, I am implementing a compiler and toolchain (Golang style) - with the semantic core in Lisette to avoid precisely this sort of problem ;-)
The semantic core is isolated as an independent stdlib-only package for now, making this a future rather than immediate  problem. 

It seems likely that `bindgen` should be bootstrapped in Lisette but there may be blocking cyclic problems or higher priorities.

The language itself is very welcome.

--- a/crates/emit/src/calls/go_interop/nullable.rs
+++ b/crates/emit/src/calls/go_interop/nullable.rs
@@ -193,6 +193,32 @@ impl Emitter<'_> {
         unwrapped_var
     }
 
+    /// Bridge `Option<T>` to Go `*T` when T is not naturally nilable.
+    pub(crate) fn emit_option_unwrap_to_go_pointer(
+        &mut self,
+        output: &mut String,
+        option_value: &str,
+        option_ty: &Type,
+    ) -> String {
+        let inner_ty = option_ty.ok_type();
+        let go_inner_ty = self.go_type_as_string(&inner_ty);
+
+        let opt_var = self.fresh_var(Some("opt"));
+        self.declare(&opt_var);
+        let ptr_var = self.fresh_var(Some("ptr"));
+        self.declare(&ptr_var);
+
+        self.flags.needs_stdlib = true;
+
+        write_line!(output, "{} := {}", opt_var, option_value);
+        write_line!(output, "var {} *{}", ptr_var, go_inner_ty);
+        write_line!(output, "if {}.Tag == {} {{", opt_var, OPTION_SOME_TAG);
+        write_line!(output, "{} = &{}.SomeVal", ptr_var, opt_var);
+        output.push_str("}\n");
+
+        ptr_var
+    }
+
     pub(crate) fn emit_collection_nullable_wrap(
         &mut self,
         output: &mut String,

--- a/crates/emit/src/expressions/access/struct_call.rs
+++ b/crates/emit/src/expressions/access/struct_call.rs
@@ -66,12 +66,18 @@ impl Emitter<'_> {
             let field_name = self.resolve_struct_call_field_name(&f.name, ty, &ctx);
             let mut value = emitted_values[fi].clone();
             value = self.wrap_recursive_enum_field(output, value, f, &ctx);
+            let value_ty = f.value.get_type();
+            let field_ty = self.lookup_struct_field_ty(ty, &f.name);
             if is_go_struct {
-                let coercion = Coercion::resolve_unwrap_go_nullable(self, &f.value.get_type());
-                value = coercion.apply(self, output, value);
+                if self.needs_go_pointer_bridge(&value_ty, field_ty.as_ref()) {
+                    value = self.emit_option_unwrap_to_go_pointer(output, &value, &value_ty);
+                } else {
+                    let coercion = Coercion::resolve_unwrap_go_nullable(self, &value_ty);
+                    value = coercion.apply(self, output, value);
+                }
             }
-            if let Some(field_ty) = self.lookup_struct_field_ty(ty, &f.name) {
-                let coercion = Coercion::resolve(self, &f.value.get_type(), &field_ty);
+            if let Some(field_ty) = field_ty.as_ref() {
+                let coercion = Coercion::resolve(self, &value_ty, field_ty);
                 value = coercion.apply(self, output, value);
             }
             field_names.push(field_name);

--- a/crates/emit/src/queries.rs
+++ b/crates/emit/src/queries.rs
@@ -244,6 +244,14 @@ impl Emitter<'_> {
         ty.is_option() && self.is_nilable_go_type(&ty.ok_type())
     }
 
+    /// True when a Go-imported struct field declared `Option<T>` (T non-nilable)
+    /// receives a value of the same shape — the field is `*T` underneath, so
+    /// the value must be bridged with address-of.
+    pub(crate) fn needs_go_pointer_bridge(&self, value_ty: &Type, field_ty: Option<&Type>) -> bool {
+        let is_non_nullable_option = |t: &Type| t.is_option() && !self.is_nullable_option(t);
+        is_non_nullable_option(value_ty) && field_ty.is_some_and(is_non_nullable_option)
+    }
+
     /// Returns true if the Option wraps a Go interface type (not a pointer).
     /// These need `IsNilInterface` instead of `!= nil` to catch typed nils.
     pub(crate) fn is_interface_option(&self, ty: &Type) -> bool {

--- a/tests/spec/emit/interop_matrix.rs
+++ b/tests/spec/emit/interop_matrix.rs
@@ -1139,3 +1139,71 @@ pub fn Peek() -> Option<Msg>
 "#;
     assert_emit_snapshot_with_go_typedefs!(input, &[("go:example.com/evts", typedef)]);
 }
+
+#[test]
+fn interop_struct_literal_option_to_pointer_named_scalar() {
+    let input = r#"
+import "go:example.com/cb"
+
+fn main() {
+  let _ = cb.Options {
+    Direction: Some(cb.DirectionDefault),
+  }
+}
+"#;
+    let typedef = r#"
+pub type Direction = int
+
+pub const DirectionDefault: Direction = 0
+
+pub struct Options {
+  pub Direction: Option<Direction>,
+}
+"#;
+    assert_emit_snapshot_with_go_typedefs!(input, &[("go:example.com/cb", typedef)]);
+}
+
+#[test]
+fn interop_struct_literal_option_to_pointer_struct() {
+    let input = r#"
+import "go:example.com/cb"
+
+fn main() {
+  let inner = cb.Inner { Tag: "x" }
+  let _ = cb.Outer {
+    Slot: Some(&inner),
+  }
+}
+"#;
+    let typedef = r#"
+pub struct Inner {
+  pub Tag: string,
+}
+
+pub struct Outer {
+  pub Slot: Option<Ref<Inner>>,
+}
+"#;
+    assert_emit_snapshot_with_go_typedefs!(input, &[("go:example.com/cb", typedef)]);
+}
+
+#[test]
+fn interop_struct_literal_option_none_to_pointer() {
+    let input = r#"
+import "go:example.com/cb"
+
+fn main() {
+  let _ = cb.Options {
+    Direction: None,
+  }
+}
+"#;
+    let typedef = r#"
+pub type Direction = int
+
+pub struct Options {
+  pub Direction: Option<Direction>,
+}
+"#;
+    assert_emit_snapshot_with_go_typedefs!(input, &[("go:example.com/cb", typedef)]);
+}

